docs/graphs/graph-concepts/graphs-and-indexes.md

```diff
@@ -0,0 +1,105 @@
+
+**Use case**: Locating posts within a specific date range or sorting edges by a weight attribute.
+
+### Persistent Index
+
+A persistent index is similar to a skiplist index in terms of supported queries, but it is stored on disk rather than in memory. This index type is useful for large datasets that do not fit in memory, but it might have slightly higher access latency compared to in-memory indexes. If your graph queries involve large datasets and require filtering or sorting based on attribute values, then a persistent index might be an appropriate choice.
+
+**Use case**: Large graphs where the index size exceeds available memory, and queries involve sorting or filtering based on attribute values.
+
+### Fulltext Index
+
+A fulltext index is designed for text-based searches, supporting word or phrase searches within the indexed attributes. If your graph queries involve searching for vertices or edges based on text content, then a fulltext index can be useful.
+
+**Use case**: Searching for posts containing specific keywords or finding users based on their interests.
+
+## Example: Using Indexes with Graphs
+
+In this theoretical example, a simple graph with two vertex collections and one edge collection is created, indexes are added to optimize query performance, and then a graph query is executed using those indexes. The graph represents a social network with users and posts, where users can create posts and follow other users.
+
+### Creating Collections
+
+Before traversing graphs, collections need to be created and data added to them:
+
+- Vertex (document) collection `users` - stores user information.
+- Vertex (document) collection `posts` - stores post information.
+- Edge collection `relations` - stores relationships between users (follows) and user-post relationships (creates).
+
+For more information about creating collections, refer to [Collections](../../collections/).
+
+### Adding Indexes
+
+After creating these collections and inserting sample data, indexes can be added to optimize query performance.
+
+For more information on adding indexes, refer to [Working with Indexes](../../collections/indexing/working-with-indexes) or the [gdnsl index](../../cli/indexes-cli) command.
+
+Add the following indexes:
+
+- **email_index**: A "hash" index on the `email` attribute in the `users` collection. This index would speed up queries that filter users based on their email addresses.
+- **from_index**: A "hash" index on the `_from` attribute in the `relations` collection. This index would optimize queries that involve traversing edges based on their starting vertices.
+
+### Running the Query
+
+With the added indexes, a graph query could be executed. Suppose the goal is to find all posts created by users followed by a specific user, identified by their email address. A C8QL query to achieve this might look like:
+
+```sql
+FOR follower IN users
```

Comment:
This multiple nested for loops is not a good example. This can be optimized using subqueries, can you provide me data set so I can do it

Reply:
`LET followers = (FOR follower IN users FILTER follower.email == 'john@example.com' RETURN follower)
LET relevantRelations = (FOR follower IN followers FOR relation IN relations FILTER relation._from == follower._id RETURN relation)
LET followees = (FOR relation IN relevantRelations FOR followee IN users FILTER followee._id == relation._to RETURN followee)
LET relevantPostRelations = (FOR followee IN followees FOR postRelation IN relations FILTER postRelation._from == followee._id RETURN postRelation)
FOR postRelation IN relevantPostRelations
  FOR post IN posts
    FILTER post._id == postRelation._to
    RETURN post
`

Reply:
In general, it's a good practice to use pre-filtering with LET statements to reduce the number of nested loops whenever possible, but this depends on the specific query and the data model.